test/system/SmokeTestHostless.cpp

```diff
@@ -0,0 +1,253 @@
+    }
+
+    const char *env_games = std::getenv("FO_TEST_HOSTLESS_GAMES");
+    if (env_games) {
+        try {
+            num_games = boost::lexical_cast<unsigned int>(env_games);
+        } catch (...) {
+            // ignore
+        }
+    }
+
+    boost::optional<Process> server;
+    if (launch_server) {
+        BOOST_REQUIRE(!PingLocalHostServer());
+
+        std::string SERVER_CLIENT_EXE = ServerClientExe();
+
+        BOOST_TEST_MESSAGE(SERVER_CLIENT_EXE);
+
+#ifdef FREEORION_MACOSX
+        // On OSX set environment variable DYLD_LIBRARY_PATH to python framework folder
+        // bundled with app, so the dynamic linker uses the bundled python library.
+        // Otherwise the dynamic linker will look for a correct python lib in system
+        // paths, and if it can't find it, throw an error and terminate!
+        // Setting environment variable here, spawned child processes will inherit it.
+        setenv("DYLD_LIBRARY_PATH", GetPythonHome().string().c_str(), 1);
+#endif
+
+        std::vector<std::string> args;
+        args.push_back("\"" + SERVER_CLIENT_EXE + "\"");
+        args.push_back("--hostless");
+        args.push_back("--save.auto.hostless.enabled");
+        args.push_back(save_game ? "1" : "0");
+        args.push_back("--testing");
+
+#ifdef FREEORION_LINUX
+        // Dirty hack to output log to console.
+        args.push_back("--log-file");
+        args.push_back("/proc/self/fd/1");
+#endif
+
+        server = Process(SERVER_CLIENT_EXE, args);
+
+        BOOST_TEST_MESSAGE("Server started.");
+    }
+
+    for (unsigned int g = 0; g < num_games; ++g) {
+        BOOST_TEST_MESSAGE(g << "Game. Connecting to server...");
+
+        BOOST_REQUIRE(ConnectToServer("localhost"));
+
+        JoinGame();
+        boost::posix_time::ptime start_time = boost::posix_time::microsec_clock::local_time();
+        start_time = boost::posix_time::microsec_clock::local_time();
+        while (!m_lobby_updated) {
+            BOOST_REQUIRE(ProcessMessages(start_time, MAX_WAITING_SEC));
+        }
+        BOOST_TEST_MESSAGE("Entered to lobby");
+
+        if (g == 0) {
+            // if first game lobby should be empty
+            BOOST_REQUIRE(GetLobbyAICount() == 0);
+
+            // fill lobby with AIs
+            for (unsigned int ai_i = 1; ai_i <= num_AIs; ++ai_i) {
+                PlayerSetupData ai_plr;
+                ai_plr.m_client_type = Networking::CLIENT_TYPE_AI_PLAYER;
+                m_lobby_data.m_players.push_back({Networking::INVALID_PLAYER_ID, ai_plr});
+                // publish changes
+                UpdateLobby();
+                start_time = boost::posix_time::microsec_clock::local_time();
+                while (!m_lobby_updated) {
+                    BOOST_REQUIRE(ProcessMessages(start_time, MAX_WAITING_SEC));
+                }
+            }
+            BOOST_REQUIRE(GetLobbyAICount() == num_AIs);
+
+        } else {
+            // other game should retain AI from previous game
```

Comment:
retain AI clients or retain number of AIs to start up?

Reply:
Number of AI clients. Fixed.